bliss/simulator/toy_simulated_dataset.py

```diff
@@ -0,0 +1,380 @@
+
+        Sample images from full catalog
+
+        If star_fluxes is of size [B,max_sources,n_bands] then we
+        expect that star_fluxes[i,j,k] = 0 whenever j>=n_sources[i]
+
+        (this facilitates the batch processing)
+        """
+
+        # make sure all of the sources are stars
+        if torch.any(full_catalog["source_type"] != 0):
+            raise ValueError("all sources must be stars")
+
+        # zero out fluxes based on n_sources
+        mask = counts_mask(full_catalog["n_sources"])[:, :, None]  # B x max_sources x 1
+        fluxes = torch.where(mask, full_catalog["star_fluxes"], 0)
+
+        # construct meshgrid of size (height,width)
+        # pixel value at img[0,0] is assumed to span from coordinate 0,0
+        # to coordinate value 1,1
+        # so the *center* of img[0,0] pixel is at coordinate (.5,.5)
+        x = torch.arange(self.width, device=full_catalog.device) + 0.5
+        y = torch.arange(self.height, device=full_catalog.device) + 0.5
+        xx, yy = torch.meshgrid(x, y, indexing="ij")
+
+        # uncollated results will have shape [B,height,width,max_sources]
+        xx = xx[None, :, :, None]
+        yy = yy[None, :, :, None]
+
+        # create B,height,width image for each band
+        band_images = []
+        for b in range(self.n_bands):
+            relevant_fluxes = fluxes[:, :, b]  # B x max_sources
+
+            img = 0
+            for coadd in range(self.coadd_depth):
+                relevant_dithers = dithers[:, coadd, b]  # B x 2
+
+                # plocs is of size [B,max_sources,2]
+                # meshgrid of size [height,width]
+                # uncollated result is of size [B,height,width,max_sources]
+                plocs_x = (
+                    full_catalog["plocs"][:, None, None, :, 0]
+                    + relevant_dithers[:, None, None, None, 0]
+                )
+                plocs_y = (
+                    full_catalog["plocs"][:, None, None, :, 1]
+                    + relevant_dithers[:, None, None, None, 1]
+                )
+
+                # compute brightness [B,height,width]
+                # we briefly store [B,height,width,max_sources] result at this
+                # stage, reaching apex memory usage
+                brightness = (
+                    torch.exp(
+                        -0.5
+                        * (
+                            (plocs_x - xx) ** 2 / self.psf_radius_h**2
+                            + (plocs_y - yy) ** 2 / self.psf_radius_w**2
+                        )  # noqa: C815
+                    )
+                    * relevant_fluxes[:, None, None]
+                ).sum(-1)
+
+                # sample and add it in
+                samp = torch.poisson(brightness + self.noise_level, generator=generator)
+                img += samp
+
+            band_images.append(img)
+
+        # merge
+        return torch.stack(band_images, dim=1)
+
+
+class ToySimulatedDataset(IterableDataset):
```

Comment:
Maybe clearer to merge this class with `ToySimulator`, for symmetry with `SimulatedDataset`.  (The reason we have two classes in `simulated_dataset.py` is that the second reads from disk.)